### PR TITLE
FIPS: add indicator for PKCS#1 version 1.5 padding usage for key transport and agreement

### PR DESCRIPTION
**Pull request by paulidale**
As mandated in [NIST SP 800-131A Rev. 2](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-131Ar2.pdf) section 6, PKCS#1 version 1.5 padding is no longer permitted for key agreement and key transport using RSA.

This is implemented here as a hard failure condition.  There is also a FIPS provider parameter that permits the use of this padding with an indicator.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: apps/fipsinstall.c
Comment (slontis) [inline apps/fipsinstall.c:98]: signing?

Comment (paulidale) [reply apps/fipsinstall.c:98]: Oops.

File: include/openssl/fips_names.h
Comment (slontis) [inline include/openssl/fips_names.h:144]: Comment?

File: providers/implementations/asymciphers/rsa_enc.c
Comment (slontis) [inline providers/implementations/asymciphers/rsa_enc.c:479]: OSSL_ASYM_CIPHER_PARAM_PKCS15_PADDING_DISABLE Needs to be in the settables list

File: include/openssl/fips_names.h
Comment (slontis) [inline include/openssl/fips_names.h:145]: Not quite right :)

File: include/openssl/fips_names.h
Comment (slontis) [inline include/openssl/fips_names.h:145]: please remove the last allowed bit

**Review by slontis — APPROVED** [review]

**Comment by slontis** [comment]
@t8m are you ok with the way the tests are done here for tls?
Disabling RSA encryption (which uses PKCS1.5), pretty much breaks a lot of TLS tests.

**Comment by paulidale** [comment]
Clean rebase to master

**Comment by paulidale** [comment]
Ping @openssl/committers for second review please.

File: apps/cms.c
Comment (beldmit) [inline apps/cms.c:1]: The changes in this file seem unrelated to the PR

Comment (paulidale) [reply apps/cms.c:1]: They are formatting fixes I did while trying to figure out why the CMS tests failed.
Happy to drop this if you want.

**Review by beldmit — APPROVED** [review]
LGTM. I'd drop the `apps/cms.c` from the PR and provide them separately but it's nt crucial

File: apps/cms.c
Comment (paulidale) [reply apps/cms.c:1]: I've dropped this.  No need for reapprovals IMO.

**Comment by t8m** [comment]
@beldmit @slontis please reconfirm as this was rebased.

**Comment by paulidale** [comment]
We'll have to rebase most of these when the merge.  They all change the exact same line in fipsprov.c 😞

**Comment by mattcaswell** [comment]
If its a trivial rebase I don't think a reconfirm is necessary.

**Comment by t8m** [comment]
AFAIK there were more changes since the approval from @slontis. However I assume @beldmit approval still holds then.

**Comment by paulidale** [comment]
This one also dropped the CMS formatting changes: three new lines and a similar number of != NULL.

**Comment by paulidale** [comment]
@beldmit @slontis ?

File: providers/implementations/asymciphers/rsa_enc.c
Comment (slontis) [inline providers/implementations/asymciphers/rsa_enc.c:164]: any reason to not use the macro here?

**Review by slontis — COMMENTED** [review]
Still approved with or without my comment

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

File: providers/implementations/asymciphers/rsa_enc.c
Comment (paulidale) [reply providers/implementations/asymciphers/rsa_enc.c:164]: Copied from somewhere else I think.  Worth fixing in due course.

Comment (slontis) [reply providers/implementations/asymciphers/rsa_enc.c:164]: FYI - It only does this in a few places in order to share the code

**Pull request closed without merging**
